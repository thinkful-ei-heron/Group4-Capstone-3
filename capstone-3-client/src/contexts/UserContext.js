import React, { Component } from 'react'
import AuthApiService from '../services/auth-api-service'
import TokenService from '../services/token-service'
import IdleService from '../services/idle-service'

const UserContext = React.createContext({
  user: {},
  language: {},
  words: [],
  error: null,
  setLanguage: () => {},
  setWords: () => {},
  setError: () => {},
  clearError: () => {},
  setUser: () => {},
  processLogin: () => {},
  processLogout: () => {},
});

export default UserContext

export class UserProvider extends Component {
  constructor(props) {
    super(props);
    const state = { user: {}, language: {}, words: [], error: null };

    const jwtPayload = TokenService.parseAuthToken();

    if (jwtPayload)
      state.user = {
        id: jwtPayload.user_id,
        name: jwtPayload.name,
        username: jwtPayload.sub,
      };

    this.state = state;
    IdleService.setIdleCallback(this.logoutBecauseIdle)
  }

  componentDidMount() {
    if (TokenService.hasAuthToken()) {
      IdleService.registerIdleTimerResets();
      TokenService.queueCallbackBeforeExpiry(() => {
        this.fetchRefreshToken()
      })
    }
  }

  componentWillUnmount() {
    IdleService.unRegisterIdleResets();
    TokenService.clearCallbackBeforeExpiry()
  }

  setError = error => {
    console.error(error);
    this.setState({ error })
  };

  clearError = () => {
    this.setState({ error: null })
  };

  setUser = user => {
    this.setState({ user })
  };

  processLogin = authToken => {
    TokenService.saveAuthToken(authToken)
    const jwtPayload = TokenService.parseAuthToken()
    this.setUser({
      id: jwtPayload.user_id,
      name: jwtPayload.name,
      username: jwtPayload.sub,
    });
  };

  processLogout = () => {
    TokenService.clearAuthToken();
    this.setUser({})
  };

  fetchRefreshToken = () => {
    AuthApiService.refreshToken()
      .then(res => {
        TokenService.saveAuthToken(res.authToken)
      })
      .catch(err => {
        this.setError(err)
      })
  };


  setLanguage = (obj) => {
    this.setState({language: obj})
  };
  setWords = (arr) => {
    this.setState({words: arr})
  };

  render() {
    const value = {
      user: this.state.user,
      error: this.state.error,
      language: this.state.language,
      words: this.state.words,
      setLanguage: this.setLanguage,
      setWords: this.setWords,
      setError: this.setError,
      clearError: this.clearError,
      setUser: this.setUser,
      processLogin: this.processLogin,
      processLogout: this.processLogout,
    };
    return (
      <UserContext.Provider value={value}>
        {this.props.children}
      </UserContext.Provider>
    )
  }
}
