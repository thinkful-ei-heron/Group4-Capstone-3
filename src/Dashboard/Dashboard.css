.journal-item-button {
  border: none;
  margin: 10px;
  padding: 10px 10px 30px 10px;
  width: 600px;
}

.journal-item-button:hover {
  cursor: pointer;
  background-color: gray;
}