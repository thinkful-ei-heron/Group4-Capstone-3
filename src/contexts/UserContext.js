import React, { Component } from 'react'
import TokenService from '../services/token-service'

const UserContext = React.createContext({
  user: {},
  error: null,
  beerList: [],
  setBeerList: () => {},
  setError: () => {},
  clearError: () => {},
  setUser: () => {},
  processLogin: () => {},
  processLogout: () => {},
});

export default UserContext

export class UserProvider extends Component {
  constructor(props) {
    super(props);
    const state = { user: {}, beerList: [], error: null };

    const jwtPayload = TokenService.parseAuthToken();

    if (jwtPayload)
      state.user = {
        id: jwtPayload.user_id,
        name: jwtPayload.name,
        username: jwtPayload.sub,
      };

    this.state = state;
  }

  setError = error => {
    console.error(error);
    this.setState({ error })
  };

  setBeerList = beerList => {
    this.setState({ beerList })
  }

  clearError = () => {
    this.setState({ error: null })
  };

  setUser = user => {
    this.setState({ user })
  };

  processLogin = authToken => {
    TokenService.saveAuthToken(authToken);

    const jwtPayload = TokenService.parseAuthToken();
    this.setUser({
      id: jwtPayload.user_id,
      name: jwtPayload.name,
      username: jwtPayload.sub,
    });

  };

  processLogout = () => {
    TokenService.clearAuthToken();
    this.setUser({})
  };

  setLanguage = (obj) => {
    this.setState({language: obj})
  };
  setWords = (arr) => {
    this.setState({words: arr})
  };

  render() {
    const value = {
      user: this.state.user,
      error: this.state.error,
      beerList: this.state.beerList,
      setBeerList: this.setBeerList,
      setError: this.setError,
      clearError: this.clearError,
      setUser: this.setUser,
      processLogin: this.processLogin,
      processLogout: this.processLogout,
    };
    return (
      <UserContext.Provider value={value}>
        {this.props.children}
      </UserContext.Provider>
    )
  }
}
