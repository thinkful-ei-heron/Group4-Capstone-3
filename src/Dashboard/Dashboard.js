import React from "react";
import NavBar from './../NavBar/NavBar';
import BeerApiService from '../services/beer-api-service';
import UserContext from '../contexts/UserContext';

class Dashboard extends React.Component {

  static contextType = UserContext

  componentDidMount() {
    this.context.clearError()
    BeerApiService.getAllBeers()
      .then(this.context.setBeerList)
      .catch(this.context.setError)
  }
  renderBeerList() {
    const { beerList = [] } = this.context
      return beerList.map
  }
    render() {
        return (
          <>
            <h2>Dashboard</h2>
            <NavBar />
            {this.renderBeerList()}
          </>
        )
    }
}

export default Dashboard;