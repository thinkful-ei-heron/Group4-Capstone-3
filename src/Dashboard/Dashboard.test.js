import React from 'react'
import ReactDOM from 'react-dom'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import UserContext, { UserProvider } from '../contexts/UserContext'

import { BrowserRouter } from 'react-router-dom';
import Dashboard from './Dashboard'

Enzyme.configure({ adapter: new Adapter() });

describe(`Dashboard component`, () => {
  it('renders without crashing', () => {
    const div = document.createElement('div')
    ReactDOM.render(
      <BrowserRouter>
        <UserProvider><Dashboard  location = {{pathname: ''}}/></UserProvider>
      </BrowserRouter>,
      div)
    ReactDOM.unmountComponentAtNode(div)
  })
  it('Displays Dashboard when rendered', () => {
    const wrapper = shallow(<UserProvider><Dashboard  location = {{pathname: ''}}/></UserProvider>)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

})