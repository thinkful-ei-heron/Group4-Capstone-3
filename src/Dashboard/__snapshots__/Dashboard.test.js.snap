// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard component Displays Dashboard when rendered 1`] = `
<ContextProvider
  value={
    Object {
      "beerList": Array [],
      "clearError": [Function],
      "error": null,
      "processLogin": [Function],
      "processLogout": [Function],
      "setBeerList": [Function],
      "setError": [Function],
      "setUser": [Function],
      "toggleExpanded": [Function],
      "user": Object {},
    }
  }
>
  <Dashboard
    location={
      Object {
        "pathname": "",
      }
    }
  />
</ContextProvider>
`;
